core/src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java

```diff
@@ -68,6 +70,35 @@ public long startRecovery(IndexShard indexShard, DiscoveryNode sourceNode,
         return status.recoveryId();
     }
 
+
+    /**
+     * Resets the recovery and performs a recovery restart on the currently recovering index shard
+     *
+     * @see IndexShard#performRecoveryRestart()
+     */
+    public void resetRecovery(long id, ShardId shardId) throws IOException {
+        try (RecoveryRef ref = getRecoverySafe(id, shardId)) {
+            // instead of adding complicated state to RecoveryTarget we just flip the
+            // target instance when we reset a recovery, that way we have only one cleanup
+            // path on the RecoveryTarget and are always within the bounds of ref-counting
+            // which is important since we verify files are on disk etc. after we have written them etc.
+            RecoveryTarget status = ref.status();
+            RecoveryTarget resetRecovery = status.resetRecovery();
+            boolean success = false;
+            try {
+                if (onGoingRecoveries.replace(id, status, resetRecovery) == false) {
+                    throw new IllegalStateException("failed to replace recovery target");
+                }
+                success = true;
+            } finally {
+                if (success == false) {
+                    onGoingRecoveries.remove(id);
```

Comment:
copying from a comment on the commit- 

nit: now that we folded the close and performRecoveryRestart into `status.resetRecovery()` we don't need the success pattern anymore. This can be:

```
if (onGoingRecoveries.replace(id, status, resetRecovery) == false) {
       resetRecovery.cancel("replace failed");
       throw new IllegalStateException("failed to replace recovery target");
}

```